subsys/fs/littlefs_fs.c

```diff
@@ -0,0 +1,730 @@
+/*
+ * Copyright (c) 2019 Bolt Innovation Management, LLC
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <kernel.h>
+#include <errno.h>
+#include <init.h>
+#include <fs.h>
+
+#define LFS_LOG_REGISTER
+#include <lfs_util.h>
+
+#include <lfs.h>
+#include <fs/littlefs.h>
+#include <flash.h>
+#include <flash_map.h>
+
+#include "fs_impl.h"
+
+struct lfs_file_data {
+	struct lfs_file file;
+	struct lfs_file_config config;
+	struct k_mem_block cache_block;
+};
+
+#define LFS_FILEP(fp) (&((struct lfs_file_data *)(fp->filep))->file)
+
+/* Global memory pool for open files and dirs */
+K_MEM_SLAB_DEFINE(file_data_pool, sizeof(struct lfs_file_data),
+		  CONFIG_FS_LITTLEFS_NUM_FILES, 4);
+K_MEM_SLAB_DEFINE(lfs_dir_pool, sizeof(struct lfs_dir),
+		  CONFIG_FS_LITTLEFS_NUM_DIRS, 4);
+
+/* If not explicitly customizing provide a default that's appropriate
+ * based on other configuration options.
+ */
+#ifndef CONFIG_FS_LITTLEFS_FC_MEM_POOL
+BUILD_ASSERT(CONFIG_FS_LITTLEFS_CACHE_SIZE >= 4);
+#define CONFIG_FS_LITTLEFS_FC_MEM_POOL_MIN_SIZE 4
+#define CONFIG_FS_LITTLEFS_FC_MEM_POOL_MAX_SIZE CONFIG_FS_LITTLEFS_CACHE_SIZE
+#define CONFIG_FS_LITTLEFS_FC_MEM_POOL_NUM_BLOCKS CONFIG_FS_LITTLEFS_NUM_FILES
+#endif
+
+K_MEM_POOL_DEFINE(file_cache_pool,
+		  CONFIG_FS_LITTLEFS_FC_MEM_POOL_MIN_SIZE,
+		  CONFIG_FS_LITTLEFS_FC_MEM_POOL_MAX_SIZE,
+		  CONFIG_FS_LITTLEFS_FC_MEM_POOL_NUM_BLOCKS, 4);
+
+static inline void fs_lock(struct fs_littlefs *fs)
+{
+	k_mutex_lock(&fs->mutex, K_FOREVER);
+}
+
+static inline void fs_unlock(struct fs_littlefs *fs)
+{
+	k_mutex_unlock(&fs->mutex);
+}
+
+static int lfs_to_errno(int error)
+{
+	if (error >= 0) {
+		return error;
+	}
+
+	switch (error) {
+	default:
+	case LFS_ERR_IO:        /* Error during device operation */
+		return -EIO;
+	case LFS_ERR_CORRUPT:   /* Corrupted */
+		return -EIO;
```

Comment:
Can this be other errorcode than -EIO? -EIO should be reserved for low level raw data access errors (like flash driver/map can't access a data block etc.).

Reply:
maybe -EFAULT?

Reply:
Switched to `-EFAULT`.  It's no more right than anything else (`LFS_ERR_CORRUPT` is used in a variety of places in littlefs, some of which map to `-EINVAL` or `-EIO`), but it does make it possible to uniquely identify the LFS-layer error which is important.